common/txmgr/tracker.go

```diff
@@ -210,53 +225,70 @@ func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) setEnabledA
 	return nil
 }
 
-// trackAbandonedTxes called once to find and insert all abandoned txes into the tracker.
+// trackAbandonedTxes called on startup to find and insert all abandoned txes into the tracker.
 func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) trackAbandonedTxes(ctx context.Context) (err error) {
-	if tr.isStarted {
-		return fmt.Errorf("tracker already started")
-	}
-
-	tr.lggr.Info("Retrieving non fatal transactions from txStore")
-	nonFatalTxes, err := tr.txStore.GetNonFatalTransactions(ctx, tr.chainID)
-	if err != nil {
-		return fmt.Errorf("failed to get non fatal txes from txStore: %w", err)
-	}
-
-	// insert abandoned txes
-	for _, tx := range nonFatalTxes {
-		if !tr.enabledAddrs[tx.FromAddress] {
-			tr.insertTx(tx)
+	return sqlutil.Batch(func(offset, limit uint) (count uint, err error) {
+		var enabledAddrs []ADDR
+		for addr := range tr.enabledAddrs {
+			enabledAddrs = append(enabledAddrs, addr)
 		}
-	}
-
-	if err := tr.handleTxesByState(ctx, 0); err != nil {
-		return fmt.Errorf("failed to handle txes by state: %w", err)
-	}
 
-	return nil
+		nonFatalTxes, err := tr.txStore.GetAbandonedTransactionsByBatch(ctx, tr.chainID, enabledAddrs, offset, limit)
+		if err != nil {
+			return 0, fmt.Errorf("failed to get non fatal txes from txStore: %w", err)
+		}
+		// insert abandoned txes
+		tr.lock.Lock()
+		for _, tx := range nonFatalTxes {
+			if !tr.enabledAddrs[tx.FromAddress] {
+				tr.txCache[tx.ID] = tx.FromAddress
+				tr.lggr.Debugf("inserted tx %v", tx.ID)
+			}
+		}
+		tr.lock.Unlock()
+		return uint(len(nonFatalTxes)), nil
+	}, batchSize)
 }
 
-func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleTxesByState(ctx context.Context, blockHeight int64) error {
+// handleTxesByState handles all txes in the txCache by their state
+// It's called on every new blockHeight and also on startup to handle all txes in the txCache
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleTxesByState(ctx context.Context, blockHeight int64) error {
 	tr.lock.Lock()
 	defer tr.lock.Unlock()
-	tr.ctx, tr.ctxCancel = context.WithTimeout(ctx, handleTxesTimeout)
-	defer tr.ctxCancel()
-	return tr.handleTxesByState(ctx, blockHeight)
-}
+	ctx, cancel := context.WithTimeout(ctx, handleTxesTimeout)
+	defer cancel()
 
-func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleTxesByState(ctx context.Context, blockHeight int64) error {
-	tr.lggr.Info("Handling transactions by state")
+	for id := range tr.txCache {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
 
-	for id, atx := range tr.txCache {
-		tx, err := tr.txStore.GetTxByID(ctx, atx.id)
+		tx, err := tr.txStore.GetTxByID(ctx, id)
 		if err != nil {
-			return fmt.Errorf("failed to get tx by ID: %w", err)
+			tr.lggr.Errorf("failed to get tx by ID: %v", err)
+			continue
+		}
+		if tx == nil {
+			tr.lggr.Warnf("tx with ID %v no longer exists, removing from tracker", id)
+			delete(tr.txCache, tx.ID)
```

Comment:
This will fail because tx is already nil:
```suggestion
			delete(tr.txCache, id)
```

Reply:
Nice catch!